### PR TITLE
[#3913, #3914] Handle enchantments that target moved fields

### PR DESCRIPTION
**Pull request by arbron**
Adds system data model for enchantment active effects that has custom application logic to handle enchantments targeting fields that no longer exist on items. These handle `attack.bonus`, `attack.flat`, `damage.parts`, `save.dc`, `save.scaling`. These fields cover everything applied by summoning as well as the most common fields modified by enchantment spells or features.

### CODE REVIEW

File: module/data/active-effect/enchantment.mjs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+            const value = damage.clone();
+            value.enchantment = true;
+            value.locked = true;
+            changes[`system.activities.${activity.id}.damage.parts`] = ActiveEffect.applyField(
+              activity, { ...change, key: "damage.parts", value: value }
+            );
+          }
+          return false;
+        } catch(err) {
+          console.log(err);
```
Comment (krbz999) [inline module/data/active-effect/enchantment.mjs:47]: Should we return false here as well if an error is thrown?

File: module/data/active-effect/enchantment.mjs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+  /**
+   * Handle enchantment-specific changes to the item.
+   * @param {Item5e} item                The Item to whom this effect should be applied.
+   * @param {EffectChangeData} change    The change data being applied.
+   * @param {Record<string, *>} changes  The aggregate update paths and their updated values.
+   * @returns {boolean|void}             Return false to prevent normal application from occurring.
+   */
+  _applyLegacy(item, change, changes) {
+    console.log("EnchantmentData#_applyLegacy", item, change, changes);
```
Comment (krbz999) [inline module/data/active-effect/enchantment.mjs:22]: Log-a-rooni.

File: module/data/active-effect/enchantment.mjs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    let key = change.key.replace("system.", "");
+    switch (change.key) {
```
Comment (Fyorl) [inline module/data/active-effect/enchantment.mjs:24]:
```suggestion
    switch ( change.key ) {
```

File: module/data/active-effect/enchantment.mjs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+          let damage;
+          const parsed = JSON.parse(change.value);
+          if ( foundry.utils.getType(parsed) === "Object" ) damage = new DamageData(parsed);
+          else damage = new DamageData({ custom: { enabled: true, formula: parsed[0][0] }, types: [parsed[0][1]] });
+          for ( const activity of item.system.activities?.getByTypes("attack", "damage", "save") ?? [] ) {
+            const value = damage.clone();
+            value.enchantment = true;
+            value.locked = true;
+            changes[`system.activities.${activity.id}.damage.parts`] = ActiveEffect.applyField(
+              activity, { ...change, key: "damage.parts", value: value }
```
Comment (Fyorl) [inline module/data/active-effect/enchantment.mjs:42]:
```suggestion
              activity, { ...change, key, value }
```

**Review by Fyorl — APPROVED** [review]
Nice

File: module/data/active-effect/enchantment.mjs (hunk 1 above)
Comment (arbron) [reply module/data/active-effect/enchantment.mjs:47]: My thought was just to fall back to the standard behavior and see if it can make sense of it.

**Pull request merged**